# Weather Text

This notebook is basically an experiment looking at historical weather text in NOAA's [Storms Events Database](https://www.ncdc.noaa.gov/stormevents/).

## Download Data

First lets download the CSV data from NOAA's [index page](https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/). You can see that they make several types of CSV files available. We are interested (for the moment) in the storm event files that go back to 1950 and have names that start like `StormEvents_details`.
[requests_html](https://requests-html.kennethreitz.org/) makes it pretty easy to scrape the CSV URLs we want out of the page, and we can load them into a huge pandas DataFrame, by concatenating each one after it has been loaded by its URL. Yes pandas can load gzipped CSV data from a URL...

In [ ]:
!pip install --quiet requests-html pandas

Get the URLs for the CSVs and download them to a `data` directory.


In [ ]:
import pandas
import requests_html

http = requests_html.HTMLSession()
url = 'https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/'
resp = http.get(url)

df = pandas.DataFrame()
for link in resp.html.find('a'):
  if 'StormEvents_details-ftp_v1.0' in link.attrs['href']:
    csv_url = url + link.attrs['href']
    df = pandas.concat([df, pandas.read_csv(csv_url)])
    print(csv_url)


https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1950_c20170120.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1951_c20160223.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1952_c20170619.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1953_c20160223.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1954_c20160223.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1955_c20160223.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1956_c20170717.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1957_c20160223.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (26,28) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1996_c20170717.csv.gz


/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (28) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1997_c20190920.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d1998_c20170717.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2000_c20200707.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2001_c20200518.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2002_c20200518.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2003_c20200518.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2004_c20200518.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2005_c20200518.csv.gz


/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (29,34,35,37) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2006_c20200518.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2007_c20170717.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2008_c20180718.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2009_c20180718.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2010_c20200922.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2011_c20180718.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2012_c20200317.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1.0_d2013_c20170519.csv.gz
https://www1.ncdc.noaa.gov/pub/data/swdi/stormevents/csvfiles/StormEvents_details-ftp_v1

In [ ]:
df

BEGIN_YEARMONTH  ...  DATA_SOURCE
0               195004  ...          PUB
1               195004  ...          PUB
2               195007  ...          PUB
3               195007  ...          PUB
4               195007  ...          PUB
...                ...  ...          ...
40032           202006  ...          CSV
40033           202006  ...          CSV
40034           202006  ...          CSV
40035           202006  ...          CSV
40036           202006  ...          CSV

[1595877 rows x 51 columns]

Since we're interested in textual content for these weather events maybe the EVENT_NARRATIVE and EPISODE_NARRATIVE columns could be interesting. Lets just try to load all the CVSs into one big DataFrame. Will Colab blow up, we'll see I guess...

Wow so it didn't blow up, it looks like it's just using 1.5GB of RAM for the 1.5 million rows?


## Narrative Text

Lets take a look at how many columns have **EPISODE_NARRATIVE**.

In [ ]:
episodes = df[df['EPISODE_NARRATIVE'].notnull()]
len(episodes) / len(all)

0.7109169441003286

So about 71% of the events have episode narrative. How about the EVENT_NARRATIVE?



In [ ]:
events = df[df['EVENT_NARRATIVE'].notnull()]
len(events) / len(all)

0.4982583244197391

Much less 50% have event narrative. About how long are these descriptions?


In [ ]:
df['EPISODE_NARRATIVE'].str.len().describe()

count   1134536.00
mean        615.10
std         932.01
min           1.00
25%         196.00
50%         351.00
75%         681.00
max       29050.00
Name: EPISODE_NARRATIVE, dtype: float64

Scientific notation makes that kinda hard to read. Let's adjust the default format for floats.

In [ ]:
pandas.set_option('display.float_format', lambda x: '%.2f' % x)

In [45]:
episodes['EVENT_NARRATIVE'].str.len().describe()

count   643153.00
mean       146.72
std        198.60
min          1.00
25%         58.00
50%         87.00
75%        154.00
max       8333.00
Name: EVENT_NARRATIVE, dtype: float64

So the longest description in here is 29050 characters long! The shortest is one character, and on average they are 615 characters long.

In [48]:
events_text_len = events['EVENT_NARRATIVE'].str.len().sum()
events_text_len

114256773

In [53]:
ulysses_len = len(http.get('https://www.gutenberg.org/files/4300/4300-0.txt').text)
ulysses_len

1586488

In [54]:
events_text_length / ulysses_len

72.01868088507446

So there are 72 Ulysses worth of weather text :-)